src/storage/src/source/postgres/replication.rs

```diff
@@ -0,0 +1,556 @@
+    let (button, errors) = builder.build_fallible(move |caps| {
+        let table_info = reader_table_info;
+        Box::pin(async move {
+            let (id, worker_id) = (config.id, config.worker_id);
+            let [data_cap, upper_cap]: &mut [_; 2] = caps.try_into().unwrap();
+            let (data_cap, upper_cap) = (data_cap.as_mut().unwrap(), upper_cap.as_mut().unwrap());
+
+            if !config.responsible_for("slot") {
+                return Ok(());
+            }
+
+            let Some(resume_lsn) = resume_upper.into_option() else {
+                return Ok(());
+            };
+            data_cap.downgrade(&resume_lsn);
+            upper_cap.downgrade(&resume_lsn);
+
+            let connection_config = connection
+                .connection
+                .config(&*context.secrets_reader)
+                .await?
+                .replication_timeouts(config.params.pg_replication_timeouts.clone());
+
+            let mut client = connection_config.connect_replication().await?;
+
+            let min_resume_lsn = super::ensure_replication_slot(&client, &connection).await?;
+            assert!(resume_lsn == MzOffset::from(0) || min_resume_lsn <= resume_lsn, "slot compacted past resumption point. resume_lsn={resume_lsn} min_resume_lsn={min_resume_lsn}");
+
+            let mut rewinds = BTreeMap::new();
+            while let Some(event) = rewind_input.next_mut().await {
+                if let AsyncEvent::Data(cap, data) = event {
+                    let cap = cap.retain_for_output(0);
+                    for req in data.drain(..) {
+                        assert!(min_resume_lsn <= req.snapshot_lsn, "slot compacted past snapshot point. snapshot_lsn={} min_resume_lsn={min_resume_lsn}", req.snapshot_lsn);
+                        rewinds.insert(req.oid, (cap.clone(), req));
+                    }
+                }
+            }
+
+            let mut committed_uppers = pin!(committed_uppers);
+
+            loop {
+                // TODO(guswynn|petrosagg): fix the underlying bug that prevents client re-use when
+                // exiting the CopyBoth mode, so we don't need to re-create clients in this function.
+                client = connection_config.connect("advance_upper").await?;
+
+                if let Err(err) = ensure_publication_exists(&client, &connection.publication).await? {
+                    // If the publication gets deleted there is nothing else to do
+                    for &oid in table_info.keys() {
+                        let update = ((oid, Err(err.clone())), *data_cap.time(), 1);
+                        data_output.give(data_cap, update).await;
+                    }
+                    return Ok(());
+                }
+
+                // We will peek the slot first to eagerly advance the frontier in case the stream
+                // is currently empty. This unblocks downstream consumers immediately instead of
+                // making them wait for TX_TIMEOUT to trigger.
```

Comment:
can you clarify this? what is a "downstream consumer"?

Reply:
I clarified the comment. Downstream consumer is any operator that consumes the output of this operator (e.g reclocking and persit_sink). This is avoiding holding back the capabilty